File: algorithms-python/sorting/CountingSort.py
'''
Counting sort ( Used when input is distributed with in some range ( 0 .. k)

Running time : Theta(n)

Fast algorithm if the input is within range.

Ex : 10 bit 10 million integers can be sorted in 5 seconds. Quick sort takes

@author: vasant
'''
import random
import datetime

def countingsort(unsorted,k):
    counter =[0 for i in range(0,k)]
    sortedArray=[0 for i in range(0,len(unsorted))]
    
    for j in range(0,len(unsorted)):
        counter[unsorted[j]]+=1
 
#rank computation
    for j in range(1,k):
        counter[j] = counter[j] + counter[j-1] 
    
    j=len(unsorted)-1
    
    while j >= 0 :
            counter[unsorted[j]]-=1   
            sortedArray[counter[unsorted[j]]] = unsorted[j]
            j-=1 
    return sortedArray

        
def printList(lst):
    for i in range(0,len(lst)):
        print lst[i]

def __main__():
    unsorted = []
    for i in range(10000000):
        unsorted.append(random.randrange(0,1024))

   # printList(unsorted)
    startTime = datetime.datetime.now()
    unsorted=countingsort(unsorted,1024) 
    endTime = datetime.datetime.now()
    print endTime - startTime
   # printList(unsorted)

if __name__ == '__main__':  
    __main__()

File: Algorithms/src/sorting/CountingSort.java
package sorting;

/**
 * Created by vasant on 2/14/16.
 */
public class CountingSort {

    public int[] sort(int[] a, int radix) {
        int[] c = new int[radix];

        for (int k = 0; k < c.length; k++) {
            c[k] = 0;
        }
        //counter
        for (int j = 0; j < a.length; j++) {
            c[a[j]]++;
        }
        //prefix sums or rank
        for (int k = 1; k < c.length; k++) {
            c[k] = c[k] + c[k - 1];
        }
        int[] b = new int[a.length];

        for (int i = 0; i < a.length; i++) {
            b[c[a[i]]] = a[i];
            c[a[i]] -= 1;
        }
        return b;
    }
}
